internal/libp2pserver/server.go

```diff
@@ -0,0 +1,94 @@
+package libp2pserver
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/filecoin-project/indexer-reference-provider/internal/p2putil"
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-msgio"
+)
+
+// Idle time before the stream is closed
+const streamIdleTimeout = 1 * time.Minute
+
+// Handler for libp2p server.
+type Handler interface {
+	HandleMessage(ctx context.Context, msgPeer peer.ID, msgbytes []byte) (proto.Message, error)
+	ProtocolID() protocol.ID
+}
+
+// Server handles client requests over libp2p
+type Server struct {
+	ctx     context.Context
+	handler Handler
+	selfID  peer.ID
+}
+
+// ID returns the peer.ID of the protocol server.
+func (s *Server) ID() peer.ID {
+	return s.selfID
+}
+
+// New creates a new libp2p Server
+func New(ctx context.Context, h host.Host, messageHandler Handler) *Server {
+	s := &Server{
+		ctx:     ctx,
+		handler: messageHandler,
+		selfID:  h.ID(),
+	}
+
+	// Set handler for each announced protocol
+	h.SetStreamHandler(messageHandler.ProtocolID(), s.handleNewStream)
+
+	return s
+}
+
+// handleNewStream implements the network.StreamHandler
+func (s *Server) handleNewStream(stream network.Stream) {
+	if s.handleNewMessages(stream) {
+		// If we exited without error, close gracefully.
+		_ = stream.Close()
+	} else {
+		// otherwise, send an error.
+		_ = stream.Reset()
+	}
+}
+
+// Returns true on orderly completion of writes (so we can Close the stream conveniently).
+func (s *Server) handleNewMessages(stream network.Stream) bool {
+	ctx := s.ctx
```

Comment:
Is this a common practice? I am curious why we don't take context as an argument.

Reply:
Unfortunately it is. Libp2p handlers can only receive a stream as input in its signature.